### PR TITLE
Add sessions to save code attempts for plugging it in

### PR DESCRIPTION
**Pull request by dhlambert**
## Proposed changes

Builds off https://github.com/uccser/cs-unplugged/pull/1400

Adds in django sessions to save a users code attempt as well as their status on the question i.e. passed, failed, started or not started.The saved statuses of each question are displayed beside the question in the navigation drawer. The saved code attempts are loaded into the editor automatically on page load if they exist.

Passed/Failed code attempts will occur once the user clicks the 'Check' button and their code is evaluated. The status will be 'Started' if the user enters in a code attempt and either opens up the navigation or clicks the Next/Previous Challenge buttons. Passed/Failed attempts cannot be overwritten by a 'Started' status. By default, each question will have a 'Not Started' status.

### Checklist

*Change the space in the box to an `x` for those you have completed.
You can also fill these out after creating the pull request.
If you're unsure about any of them, don't hesitate to ask.
We're here to help!
This is simply a reminder of what we are going to look for before merging your change.*

- [ ] I have read the contribution guidelines.
- [ ] I have linked any relevant existing issues/suggestions in the description above (include `#???` in your description to reference an issue, where `???` is the issue number).
- [X] The pull request is requesting a merge into the `develop` branch.
- [ ] I have added necessary documentation (if appropriate).
- [ ] If I've added/modified/deleted a third-party system, I've updated the relevant license files.

### Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc.
Feel free to add any images that might be helpful to understand the initial problem/solution.

### CODE REVIEW

File: csunplugged/static/js/editor-options-menu.js [hunk 1]
```diff
@@ -11,66 +11,71 @@ function openNav() {
...
+
+  return '#'
 }
 
 /**
  * Sets the link to the previous challenge if it is not the first challenge.
- * @param {Number} index The index of the current challenge
  */
-function setPreviousChallenge(index) {
+function getPreviousChallengeURL() {
```
[the remark this answers is not in the document]
Comment (dhlambert) [reply csunplugged/static/js/editor-options-menu.js:47]: Good idea

File: csunplugged/static/js/jobe-editor.js [hunk 2]
```diff
@@ -103,9 +155,16 @@ function downloadCode() {
...
-let downloadButton = document.getElementById("download_button");
-downloadButton.addEventListener("click", downloadCode);
+$("#download_button").click(downloadCode);
+
+// Apply the navigation setup
+editorUtils.setupLessonNav()
+
+// Save code when the user navigates using the next or prev buttons or opening nav
+// Manually navigating to the next page to ensure code is saved first before the page reloads.
+$("#prev_challenge_button").on("click", () => save_code().then(() => window.location.href = editorUtils.getPreviousChallengeURL()));
```
[the remark this answers is not in the document]
Comment (dhlambert) [reply csunplugged/static/js/jobe-editor.js:168]: I can't seem to replicate the first error either, but I think I've fixed the second error now.

File: csunplugged/static/js/editor-options-menu.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (dhlambert) [reply csunplugged/static/js/editor-options-menu.js:47]: The buttons are now hidden if there is not another challenge in that direction, as suggested by Jack.

File: csunplugged/static/scss/programming-editor.scss [hunk 3]
```diff
@@ -371,20 +371,3 @@ body {
...
-.status-badge-incorrect {
-  border-color: #cc0423;
-  background-color: #cc0423;
-  color: white;
-}
-
-.status-badge-started {
-  background-color: #ffc107;
-  border-color: #ffc107;
-}
```
Comment (courtneycb) [inline csunplugged/static/scss/programming-editor.scss]: no newline at EOF

File: csunplugged/templates/plugging_it_in/programming-challenge.html [hunk 4]
```diff
@@ -56,14 +56,14 @@ <h5>{{ programming_challenge_item.challenge_set_number }}.{{ programming_challen
               <div class='col-2 d-flex align-items-center justify-content-center'>
               {% if programming_challenge_item.slug in saved_attempts %}
                 {% if saved_attempts|get_item:programming_challenge_item.slug|get_item:"status" == "failed" %}
-                  <p class="status-badge status-badge-incorrect">Incorrect</p> 
+                  <p class="status-badge bg-danger border-danger text-white">{% trans "Incorrect" %}</p>
```
Comment (courtneycb) [inline csunplugged/templates/plugging_it_in/programming-challenge.html:59]: _Might_ not need the `text-white`? Try it and see?

Comment (dhlambert) [reply csunplugged/templates/plugging_it_in/programming-challenge.html:59]: I tried it earlier it kept it as the darker text, so think it might be needed.

**Comment by dhlambert** [comment]
I've added a test for the custom dictionary tag. Hopefully it will pass the codecov now.

File: csunplugged/tests/general/templatetags/test_custom_tags.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+            "Hello",
+            rendered
+        )
+
+    @override_settings(LANGUAGES=AVAILABLE_LANGUAGES_EN)
+    def test_get_nonexistant_dict_value(self):
+        context = {
+            "a_dictionary": {"test_key": "test_value"}
+        }
+        # tag = "{% load custom_tags %}{% if a_dictionary|get_item:test_key != 'test_value' %}Hello{% endif %}"
```
Comment (courtneycb) [inline csunplugged/tests/general/templatetags/test_custom_tags.py:35]: comment left in?

**Pull request merged**